numpy/core/tests/test_numeric.py

```diff
@@ -1306,6 +1327,202 @@ def test_nonzero_twodim(self):
         assert_equal(np.nonzero(x['a'].T), ([0, 1, 1, 2], [1, 1, 2, 0]))
         assert_equal(np.nonzero(x['b'].T), ([0, 0, 1, 2, 2], [0, 1, 2, 0, 2]))
 
+        types = [bool, np.int8, np.int16, np.int32, np.int64, np.uint8, np.uint16, np.uint32, np.uint64, np.float32, np.float64]
+        idxs = np.arange(100)
+        iters = 10
+        dim_permutation = np.arange(2) # Shuffle dimensions
+
+        for _ in range(iters):
+            for dtype in types:
+                x = ((2**33)*np.random.randn(10, 10)).astype(dtype)
+                np.random.shuffle(idxs)
+                num_zeros = np.random.randint(0, 101)
+                x.flat[idxs[:num_zeros]] = 0
+                idxs_0, idxs_1 = np.nonzero(x)
+                assert_equal(np.count_nonzero(x), len(idxs_0))
+                nzs = 0
+                for i,j in zip(idxs_0, idxs_1):
+                    if x[i,j] == 0:
+                        nzs += 1
+                assert_equal(nzs, 0)
+
+
+        for _ in range(iters): # This time randomly permute dimensions
+            for dtype in types:
+                x = ((2**33)*np.random.randn(10, 10)).astype(dtype)
+                np.random.shuffle(idxs)
+                num_zeros = np.random.randint(0, 101)
+                x.flat[idxs[:num_zeros]] = 0
+                np.random.shuffle(dim_permutation)
+                x = np.transpose(x, dim_permutation)
+                idxs_0, idxs_1 = np.nonzero(x)
+                assert_equal(np.count_nonzero(x), len(idxs_0))
+                nzs = 0
+                for i,j in zip(idxs_0, idxs_1):
+                    if x[i,j] == 0:
+                        nzs += 1
+                assert_equal(nzs, 0)
```

Comment:
Its fine to implement nonzero in Python here, but please refactor it. You can use `np.nditer` if that helps.  Or flatten and `np.unravel_index` (and `np.ravel()`) if the n dimensions are annoying to deal with.